src/library/scala/collection/immutable/ChampHashMap.scala

```diff
@@ -849,13 +860,34 @@ class MapBuilder[K, V] extends Builder[(K, V), HashMap[K, V]] {
     bm.size = bm.size - 1 + node.size
   }
 
-  def setValue[V1 >: V](bm: BitmapIndexedMapNode[K, V], bitpos: Int, newKey: K, newValue: V1): Unit = {
+  /** Mutates `bm` to replace inline data at bit position `bitpos` with updated key/value */
+  private def setValue[V1 >: V](bm: BitmapIndexedMapNode[K, V], bitpos: Int, newKey: K, newValue: V1): Unit = {
     val dataIx = bm.dataIndex(bitpos)
     val idx = TupleLength * dataIx
     bm.content(idx) = newKey
     bm.content(idx + 1) = newValue
   }
 
+  private def insertIfNotExists(v: Vector[(K, V)], key: K, keyHash: Int, value: V): Vector[(K, V)] = {
+    var i = 0
+    val len = v.length
+    while (i < len) {
+      val elem = v(i)
+      if (elem._1 == key) {
+        if (elem._2 == value) {
+          return v
```

Comment:
We should not be calling `==` on values, other than in `Map.equals`.

Furthermore, we expect that in:

```
assert(k1 == k2 && (k1 ne ke))
val map = builder.addOne(k1, "").addOne(k2, "").result()
assert(map.find(_._1 == k2).get._1 eq k2)
```

So this optimization should be restricted to the case when both `key` and `value` are `eq` to the existing binding. @mkeskells's proposal to change collision nodes internal representation from `Vector` to `Array` would mean we could just do an array mutation.

Reply:
@retronym thank you for the feedbck 👍 
>@mkeskells's proposal to change collision nodes internal representation from Vector to Array would mean we could just do an array mutation.  
  
Should we leave the Vector -> Array change for an other PR, or should we fold that into this one?

Reply:
It currently lives in #7117. You could cherry-pick it here to see how far we could go to push performance.